addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/ui/TabbedOutputPanel.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2024 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.commonlib.ui;
+
+import java.awt.BorderLayout;
+import java.awt.Component;
+import java.awt.event.KeyEvent;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.swing.AbstractButton;
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.JButton;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.JToolBar;
+import javax.swing.text.DefaultCaret;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.extension.AbstractPanel;
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.view.OutputPanel;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.extension.help.ExtensionHelp;
+import org.zaproxy.zap.utils.DisplayUtils;
+import org.zaproxy.zap.utils.ThreadUtils;
+import org.zaproxy.zap.utils.TimeStampUtils;
+import org.zaproxy.zap.utils.ZapTextArea;
+import org.zaproxy.zap.view.OutputSource;
+import org.zaproxy.zap.view.TabbedPanel2;
+import org.zaproxy.zap.view.ZapToggleButton;
+
+/**
+ * A tabbed version of the output panel that allows multiple sources of output to be displayed in
+ * separate tabs.
+ *
+ * @since 1.30.0
+ */
+@SuppressWarnings("serial")
+public class TabbedOutputPanel extends OutputPanel {
+
+    public static final String ATTRIBUTE_ICON = "commonlib.output.panel.icon";
+    public static final String ATTRIBUTE_ADDITIONAL_BUTTONS =
+            "commonlib.output.panel.additionalButtons";
+
+    private static final String DEFAULT_OUTPUT_SOURCE_NAME =
+            Constant.messages.getString("commonlib.output.panel.default");
+    private static final String ERROR_OUTPUT_SOURCE_NAME =
+            Constant.messages.getString("commonlib.output.panel.error");
+
+    private static final String CLEAR_BUTTON_TOOL_TIP =
+            Constant.messages.getString("commonlib.output.panel.button.clear.toolTip");
+
+    private static final ImageIcon DOC_ICON = getImageIcon("/resource/icon/16/172.png");
```

Comment:
Rename the file so that it's meaningful? (I'm guessing that you've copied this from core or from the fuge site, but for my 2 cents meaningful names are much more helpful.)

Reply:
This is using the core resource directly. I didn't copy it to this add-on but I can do that if that's preferable.

Reply:
That's fine. Leave it for now.